x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/StreamingHttpResultPublisher.java

```diff
@@ -0,0 +1,244 @@
+ * the HttpEntity.</p>
+ */
+class StreamingHttpResultPublisher implements HttpAsyncResponseConsumer<HttpResponse>, Flow.Publisher<HttpResult> {
+    private final HttpSettings settings;
+    private final ActionListener<Flow.Publisher<HttpResult>> listener;
+
+    // used to manage the HTTP response
+    private volatile HttpResponse response;
+    private volatile Exception ex;
+
+    // used to control the state of this publisher (Apache) and its interaction with its subscriber
+    private final Deque<HttpResult> resultQueue = new ConcurrentLinkedDeque<>();
+    private final PendingRequestCount requestCount = new PendingRequestCount();
+    private final AtomicBoolean isDone = new AtomicBoolean(false);
+    private final AtomicBoolean subscriptionCanceled = new AtomicBoolean(false);
+    private volatile Flow.Subscriber<? super HttpResult> subscriber;
+
+    private final RequestBasedTaskRunner taskRunner;
+
+    // used to control the flow of data from the Apache client, if we're producing more bytes than we can consume then we'll pause
+    private final AtomicLong totalByteSize = new AtomicLong(0);
+    private final Object ioLock = new Object();
+    private volatile IOControl savedIoControl;
+
+    StreamingHttpResultPublisher(ThreadPool threadPool, HttpSettings settings, ActionListener<Flow.Publisher<HttpResult>> listener) {
+        this.settings = Objects.requireNonNull(settings);
+        this.listener = Objects.requireNonNull(listener);
+
+        this.taskRunner = new RequestBasedTaskRunner(new OffloadThread(), threadPool, UTILITY_THREAD_POOL_NAME);
+    }
+
+    @Override
+    public void responseReceived(HttpResponse httpResponse) throws IOException {
+        this.response = httpResponse;
+        this.resultQueue.offer(HttpResult.create(settings.getMaxResponseSize(), response));
+        this.listener.onResponse(this);
+    }
+
+    @Override
+    public void subscribe(Flow.Subscriber<? super HttpResult> subscriber) {
+        this.subscriber = subscriber;
+        subscriber.onSubscribe(new HttpSubscription());
+    }
+
+    @Override
+    public void consumeContent(ContentDecoder contentDecoder, IOControl ioControl) throws IOException {
+        // if the subscriber canceled us, tell Apache
+        if (subscriptionCanceled.get()) {
+            ioControl.shutdown();
+            return;
+        }
+
+        var buffer = new SimpleInputBuffer(4096);
+        var consumed = buffer.consumeContent(contentDecoder);
+        var allBytes = new byte[consumed];
+        buffer.read(allBytes);
+
+        // we can have empty bytes, don't bother sending them
+        if (allBytes.length > 0) {
+            resultQueue.offer(new HttpResult(response, allBytes));
+        }
+
+        // always check if totalByteSize > the configured setting in case the settings change
+        if (totalByteSize.accumulateAndGet(allBytes.length, Long::sum) >= settings.getMaxResponseSize().getBytes()) {
```

Comment:
If `settings.getMaxResponseSize()` is exceeded then the input is paused until some of the data has been processed. is that correct?

`settings.getMaxResponseSize()` defaults to 50MB, should we stop and quit entirely at that point? If the intention is slow down the incoming response maybe check the sizeof `resultQueue` to see if processing is keeping up.

Reply:
Ah, I meant to decrement `totalByteSize` once we poll from `resultQueue` :frowning_face: